### PR TITLE
Add trashbin expiration cronjob

### PR DESCRIPTION
**Pull request by VicDeo**
Implement a background job that expires trashbin items.
Ref #16300

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
@MorrisJobke @Xenopathic @blizzz @PVince81 please review and test - THX

**Comment by MorrisJobke** [comment]
Tested, works and code looks good :+1:

File: apps/files_trashbin/lib/backgroundjob/expiretrash.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	/**
+	 * Act on behalf on trash item owner
+	 * @param string $user
+	 * @return boolean
+	 */
+	private function setupFS($user){
+		$userManager = \OC::$server->getUserManager();
+		if (!$userManager->userExists($user)) {
```
Comment (PVince81) [inline apps/files_trashbin/lib/backgroundjob/expiretrash.php]: Nice!

**Comment by PVince81** [comment]
Tested with the retention setting. After running cron it removes the entries from trash.

While the entries got remove from `oc_files_trash`, the files still exist on-disk. (could be a separate bug)

**Comment by PVince81** [comment]
My steps:

1. Set retention setting to "1, 2" in config.php
1. Delete a bunch of files inside a folder
1. SQL: `update oc_files_trash set timestamp=timestamp-3600*48;` (to make older)
1. SQL: `update oc_jobs set last_run=0;`
1. CLI: `sudo -uwwwrun php cron.php`
1. Check the oc_files_trash table: empty
1. Check the web UI trashbin: files are still there.

@MorrisJobke how did you test ?

**Comment by MorrisJobke** [comment]
* set the retention to `auto, 30`
* set interval to 1 (second) https://github.com/owncloud/core/blob/trashbin-expiration-cronjob/apps/files_trashbin/lib/backgroundjob/expiretrash.php#L45
* set maxAge to seconds by removing the `86400` in https://github.com/owncloud/core/blob/trashbin-expiration-cronjob/apps/files_trashbin/lib/expiration.php#L112-112
* remove the file
* run cronjob
* file is still there -> also in the webui and in the data folder
* wait 30 seconds
* run cronjob
* file is gone.

**Comment by PVince81** [comment]
@VicDeo can you have a look at my test case ?

**Comment by VicDeo** [comment]
@PVince81 I suspect that you need to change filenames under `data/user/files_trashbin/files` to have the same timestamps  as `update oc_files_trash set timestamp=timestamp-3600*48;`
because the deletion time is  stored there too.

cc @schiesbn

**Comment by schiesbn** [comment]
I would not base the expire operation on the database. We can't guarantee that the oc_trashbin table is always in sync the the files in the trashbin. In the past we saw that people regularly delete files from the trashbin manually on their servers. That's the reasons why we only use the table to remember where the files come from but we don't require it. If no entry exists the file will simply restored to the root.

Therefore I would recommend two expire jobs:

1. Expire files from the trash bin on the file system level. Just iterate over the root of user/files_trashbin (you don't have to check the sub folders). The filename contains everything you need, including the timestamp. 

2. A second cronjob checks the oc_trashbin table if the files stored in the table still exists and clean up the table. This cronjob can also run independent from the first one just every day, week or even month. Nothing should break if some entries exists for a few days after the real file was deleted before it get cleaned up.

**Comment by VicDeo** [comment]
@schiesbn I don't see a reason for maintaining a DB table in this case. It looks like `Let's create a table just to to keep it in sync with FS by means of a cronjob that we should create later`

**Comment by schiesbn** [comment]
@VicDeo The purpose of the table is to remember where the files comes from so that it can be restored to the original location. But it is not a hard requirement, if the entry doesn't exists we just fall back to the root

**Comment by VicDeo** [comment]
I just would like to note that there is  HUGE performance difference between
 iterating through all filesystem items in the folders of different users to find items older than a given timestamp
and 
executing one SELECT for all users which this timestamp inside SQL query

**Comment by schiesbn** [comment]
Sure, this way you have one select for each user. But I think that's OK because it happens in a background job.

**Comment by karlitschek** [comment]
I think it is ok because it is a background job, as @schiesbn said.
Who can help to review this?
:+1: from me

**Comment by schiesbn** [comment]
@VicDeo what is the state? Did you consider adding the changes I suggested?

**Comment by cmonteroluque** [comment]
@VicDeo ping

**Comment by VicDeo** [comment]
@cmonteroluque pong.
Alternative implementation not using DB submitted as https://github.com/owncloud/core/issues/18943
choose any appropriate
cc  @schiesbn

**Comment by schiesbn** [comment]
@VicDeo thanks for providing the alternative aproach!

I clearly vote for  #18943 

:-1: for this one.

**Comment by VicDeo** [comment]
closed in favor of https://github.com/owncloud/core/pull/18943

**Comment by cmonteroluque** [comment]
@VicDeo thanks!

**Pull request closed without merging**